internal/heartbeat/heartbeat.go

```diff
@@ -0,0 +1,119 @@
+package heartbeat
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/contentsquare/chproxy/config"
+)
+
+var ErrUnexpectedResponse = fmt.Errorf("unexpected response")
+
+type HeartBeat interface {
+	IsHealthy(ctx context.Context, addr string) error
+	Interval() time.Duration
+}
+
+type heartBeatOpts struct {
+	defaultUser	 string
+	defaultPassword string
+}
+
+type Option interface {
+	apply(*heartBeatOpts)
+}
+
+type defaultUser struct {
+	defaultUser	 string
+	defaultPassword string
+}
+
+func (o defaultUser) apply(opts *heartBeatOpts) {
+	opts.defaultUser = o.defaultUser
+	opts.defaultPassword = o.defaultPassword
+}
+
+func WithDefaultUser(user, password string) Option {
+	return defaultUser{
+		defaultUser: user,
+		defaultPassword: password,
+	}
+}
+
+type heartBeat struct {
+	interval time.Duration
+	timeout  time.Duration
+	request  string
+	response string
+	user     string
+	password string
+}
+
+// User credentials are not needed
+const DefaultEndpoint string = "/ping"
+
+func NewHeartbeat(c config.HeartBeat, options ...Option) HeartBeat {
+	opts := &heartBeatOpts{}
+	for _, o := range options {
+		o.apply(opts)
+	}
+
+	newHB := &heartBeat{
+		interval: time.Duration(c.Interval),
+		timeout:  time.Duration(c.Timeout),
+		request:  c.Request,
+		response: c.Response,
+	}
+
+	if c.Request != DefaultEndpoint {
+		if c.User != "" {
+			newHB.user = c.User
+			newHB.password = c.Password
+		} else {
+			newHB.user = opts.defaultUser
+			newHB.password = opts.defaultPassword
+		}
+	}
+
+	return newHB
+}
+
+func (hb *heartBeat) IsHealthy(ctx context.Context, addr string) error {
+	req, err := http.NewRequest("GET", addr+hb.request, nil)
+	if err != nil {
+		return err
+	}
+	if hb.request != DefaultEndpoint && hb.user != "" {
```

Comment:
although it was in the previous code, the `hb.user != ""` seems useless because this condition is not possible at this point of the code since the NewHeartbeat use a default user in the worst case scenario and this default user must be not empty cf config.go:L461 & proxy.go:L669 .

If we don't want to rely on business code outside of this package  (although the test coverage should spot this issue), it might be cleaner to do a check in the `func NewHeartbeat(c config.HeartBeat, options ...Option) HeartBeat` that there is at least a user not empty (and return an error otherwise). Because, with the current logic, if there is a regression outside of the package, it might result of have the authentication not working.

Reply:
That is a good point, I included a check in NewHeartbeat to make sure that authentication can't be broken by forgetting to pass a default user.